pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -82,37 +79,40 @@ public class ModelInterpreter implements Serializable {
-                                                        setUpDelegate(thisStage.steps.closure).call()
-                                                    }.call()
-                                                } catch (Exception e) {
-                                                    script.echo "Error in stages execution: ${e.getMessage()}"
-                                                    script.getProperty("currentBuild").result = Result.FAILURE
-                                                    if (firstError == null) {
-                                                        firstError = e
-                                                    }
-                                                } finally {
-                                                    // And finally, run the post stage steps.
-                                                    List<Closure> postClosures = thisStage.satisfiedPostStageConditions(root, script.getProperty("currentBuild"))
+                                script.stage(thisStage.name) {
+                                    withEnvBlock(thisStage.getEnvVars()) {
+                                        if (firstError == null) {
+                                            inDeclarativeAgent(thisStage.agent) {
+                                                toolsBlock(thisStage.agent ?: root.agent, thisStage.tools) {
+                                                    try {
+                                                        catchRequiredContextForNode(root.agent) {
+                                                            setUpDelegate(thisStage.steps.closure).call()
+                                                        }.call()
+                                                    } catch (Exception e) {
+                                                        script.echo "Error in stages execution: ${e.getMessage()}"
+                                                        script.getProperty("currentBuild").result = Result.FAILURE
+                                                        if (firstError == null) {
+                                                            firstError = e
+                                                        }
+                                                    } finally {
+                                                        // And finally, run the post stage steps.
+                                                        List<Closure> postClosures = thisStage.satisfiedPostStageConditions(root, script.getProperty("currentBuild"))
                                                         catchRequiredContextForNode(thisStage.agent != null ? thisStage.agent : root.agent, false) {
                                                             if (postClosures.size() > 0) {
                                                                 script.echo("Post stage")
```

Comment:
Ideally, yes, but until Blue Ocean is comfortable with nested stages, we're avoiding them.